docs/cookbook.md

```diff
@@ -0,0 +1,87 @@
+# Odoyle's Rules Cookbook
+
+## Initializing values
+
+Suppose we have an aggregate:
+
+```clojure
+(o/ruleset
+  {#_...
+   ::agregate-things
+   [:what 
+    [thing-id :thing/name  thing-name]
+    [thing-id :thing/owner person-id]
+    :then-finally
+    (->> (o/rules o/*session* ::aggregate-things)
+         (group-by :thing-id)
+         (o/insert o/*session* :global :domain/things)
+         (o/reset!))]})
+```
+
+The trouble is that `:thing/owner` may be an attribute of some things, but not all of them. We want to aggregate all the things.
+
+The solution is to use a rule that initializes a `nil` value for `:thing/owner`:
+
+```clojure
+(o/ruleset
+  {#_...
+   ::initialize-thing-owner
+   [:what [_ :thing/id id {:then not=}]
+    :then (o/insert! id :thing/owner nil)]})
+```
+
+The value may not always be nil. It may also often be 0 or another neutral value. Note that `{:then not=}` prevents the rule from firing more than once. (However, if `[id :thing/id id]` is used, this will not work.)
+
+## Loading Data 
+
+Instead of gathering the data required for your rules, shaping it into the form you need it, and inserting it with `o/insert`, you can use rules to make the process automatic.
+
+For example:
+
+```clojure
+(ns app.rules
+  (:require [odoyle.rules :as o]
+            [app.thing.db :refer [fetch-thing]])) 
+            ;; fetch-thing queries the db
+            ;; which returns a namespaced map
+                                                   
+(def rules
+  (o/ruleset 
+    {::fetch-thing 
+     [:what [_ :thing/id id {:then not=}]
+     [:then (o/insert! id (fetch-thing id))]]}))
+```
+
+Now, instead of calling `fetch-thing` and then `o/insert` from the outside, you can use `(o/insert <some-id> :thing/id <some-id>)`, and the rule will fetch for you.
+
+The power of this approach is that O'Doyle rules becomes a query engine.
+
+Suppose that you have `thing/owner` attribute, which is a set of the things a person owns. Combining two fetchers reveals the power of this approach:
+
+```clojure
+(ns app.rules
+  (:require [odoyle.rules :as o]
+            [app.person.db :refer [fetch-person]]
+            [app.thing.db :refer [fetch-thing]])) 
+            
+(def rules
+  (o/ruleset 
+    {::fetch-thing
+     [:what [_ :thing/id id {:then not=}]
+      :then                                     ;; fetch-thing returns a namespaced map
+      (o/insert! id (fetch-thing id))]          ;; `{:thing/name "name", :thing/owner 123, #_...}`
```

Comment:
In this example, is the `fetch-thing` fn getting the data from the session via `o/query-all`, or from somewhere else?

Reply:
In this example, `fetch-thing` is returning a map from the database, not from the session. (The map here is flattened and namespaced.)

The benefit of this in my use cases is that I can put a single id for an object, and the rules will fetch all the related objects. This saves me from having to gather everything up before hand, which can often span quite a few related entities and domains.

Reply:
Yeah that makes sense. Is there any benefit to doing the db query from inside a rule, vs querying and inserting from the outside? I tend to avoid interacting with external things (databases, rendering, etc) from inside rules, though this is probably OK. One big gotcha that I mention on the readme is using `query-all` from within a rule that queries another rule. That could cause problems because it eliminates reactivity. But if you're just querying some external db then it won't be reactive anyway.

Reply:
Putting the queries in the rules handles the joins for you automatically, somewhat like what [Pathom](https://pathom3.wsscode.com/docs/resolvers/) does. If I insert one id for an entity, the rules engine will make all the queries, find all the relations, load all the related entities, and execute the rules against them all.

I've found this to be extremely useful.

To give a concrete example: at work, we have a library of rules in microservice environment. If I'm in the service that owns an entity, I want to pass in a db connection so I can use transactions. But an entity in a service is almost always related to entities in another service. Those entities I want to fetch over a network call to another service (or from Redis, depending on what it is). 

Again, I might be using the library from the front end, in which case I want to make a network call for entities, but need to make those calls differently (authorization rules).

I can pass a triple to the rules engine indicating which service the rules engine is being used from. Then the rules govern how an entity is loaded based on the fact indicating where it is being called from. These are written once for all services, but handle all the different use cases.

So the rules engine turns into a pretty powerful query system + rules engine. For my use case, the query part is as much part of the value proposition as the "normal usage" of the rules.

(One potential drawback is that all these network calls need to be blocking. We typically block on network calls anyway, so this hasn't been an issue.)

Reply:
In the case of `::fetch-thing` there is no join. When you use this technique are you normally querying with more than just an id, so a join is needed? Just want to understand the use case.